xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,725 @@
+        @Override
+        public void run() {
+          handleStreamClosed(
+              Status.fromThrowable(t).augmentDescription("ADS stream [" + this + "] had an error"));
+        }
+      });
+    }
+
+    @Override
+    public void onCompleted() {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(
+              Status.UNAVAILABLE.withDescription("ADS stream [" + this + "] was closed"));
+        }
+      });
+    }
+
+    private void handleStreamClosed(Status error) {
+      checkArgument(!error.isOk(), "unexpected OK status");
+      if (closed) {
+        return;
+      }
+      closed = true;
+      cleanUp();
+      configWatcher.onError(error);
+      if (responseReceived || retryBackoffPolicy == null) {
+        // Reset the backoff sequence if had received a response, or backoff sequence
+        // has never been initialized.
+        retryBackoffPolicy = backoffPolicyProvider.get();
+      }
+      long delayNanos = 0;
+      if (!responseReceived) {
+        delayNanos =
+            Math.max(
+                0,
+                retryBackoffPolicy.nextBackoffNanos() - stopwatch.elapsed(TimeUnit.NANOSECONDS));
+      }
+      logger.log(Level.FINE, "{0} stream closed, retry in {1} ns", new Object[]{this, delayNanos});
+      if (delayNanos == 0) {
+        startRpcStream();
+        if (configWatcher != null) {
+          adsStream.sendXdsRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName));
+        }
+        // TODO(chengyuanzhang): send CDS/EDS requests if CDS/EDS watcher presents.
+      } else {
+        rpcRetryTimer =
+            syncContext.schedule(
+                new RpcRetryTask(), delayNanos, TimeUnit.NANOSECONDS, timeService);
+      }
+    }
+
+    private void close(Exception error) {
+      if (closed) {
+        return;
+      }
+      closed = true;
+      cleanUp();
+      requestWriter.onError(error);
+    }
+
+    private void cleanUp() {
+      if (adsStream == this) {
+        adsStream = null;
+      }
+    }
+
+    /**
+     * Sends a DiscoveryRequest for the given resource name to management server. Memories the
+     * requested resource name (except for LDS as we always request for the singleton Listener)
+     * as we need it to find resources in responses.
+     */
+    private void sendXdsRequest(String typeUrl, Collection<String> resourceNames) {
+      checkState(requestWriter != null, "ADS stream has not been started");
+      String version = "";
+      String nonce = "";
+      if (typeUrl.equals(ADS_TYPE_URL_LDS)) {
+        version = ldsVersion;
+        nonce = ldsRespNonce;
+      } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+        checkArgument(resourceNames.size() == 1,
+            "RDS request requesting for more than one resource");
+        version = rdsVersion;
+        nonce = rdsRespNonce;
+        rdsResourceName = resourceNames.iterator().next();
```

Comment:
For the RDS case I noticed redundancy and unnecessary complexity: in all (but one) cases you are passing `this.rdsResourceName` via the `resourceNames` parameter (as `ImmutableList.of`) and then using the same parameter value to set  `this.rdsResourceName`. This is unnecessarily complex and should be eliminated - why pass a member variable as a param? And then set it back as that member variable?

Reply:
I inspected all usages of `sendXdsRequest(...)` method, did not find the usage you mentioned. Could you point it out the specific one? This method takes in the collection of resource names, memories them in `AdsStream` instance and sends out a request.

Reply:
Okay, I see 4 calls to `sendXdsRequest` now (I think there were more in earlier versions) and when it is for `ADS_TYPE_URL_LDS` you are always passing `ImmutableList.of(ldsResourceName)` where `ldsResourceName` is a member of `XdsClientImpl` and when it is for `ADS_TYPE_URL_RDS` (only 1 call) it is (line 333):
```
adsStream.sendXdsRequest(ADS_TYPE_URL_RDS, ImmutableList.of(rdsRouteConfigName));
```
but `rdsRouteConfigName` is already assigned to `AdsStream.rdsResourceName` on line 322 earlier. So I am saying in all of these usages the resource name(s) is always the member variable (either XdsClientImpl.ldsResourceName or AdsStream.rdsResourceName) so there is no need to pass it as a parameter. However in future you might have more cases where you really need to pass resource names so you can decide if this comment applies or not.

Reply:
Oh, you are saying the case for line 322. That's different, and it has special meanings. Please see https://github.com/grpc/grpc-java/pull/6320#discussion_r347066305 (line 324 now is 322).